README.md
[![CodeFactor](https://www.codefactor.io/repository/github/rogerselwyn/skyq_remote/badge)](https://www.codefactor.io/repository/github/rogerselwyn/skyq_remote)

![GitHub release](https://img.shields.io/github/v/release/RogerSelwyn/skyq_remote) [![maintained](https://img.shields.io/maintenance/yes/2023.svg)](#)
[![maintainer](https://img.shields.io/badge/maintainer-%20%40RogerSelwyn-blue.svg)](https://github.com/RogerSelwyn)
[![Community Forum](https://img.shields.io/badge/community-forum-brightgreen.svg)](https://community.home-assistant.io/t/custom-component-skyq-media-player/140306)

# pyskyqremote
Python module for accessing SkyQ box and EPG, and sending commands

## Introduction

This library enables access to SkyQ devices, primarily focused on the UK, but should also work for Italy and Germany.

## Installing

To install:

```
pip install pyskyqremote
```

## Usage

### Base
```
from pyskyqremote.skyq_remote import SkyQRemote

self.client = SkyQRemote('192.168.1.99')
```
Optional parameters:
* epg_cache_len - Default = 20
* port - Default = 49160
* json_port - Default = 9006


### Get device information

```
device = self.client.get_device_information()
```

Will return an object such as below for device informatiom:

```
{
   'ASVersion':'Q112.000.21.00-AS_asdev',
   'IPAddress':'192.168.1.22',
   'countryCode':'GBR',
   'usedCountryCode':'GBR',
   'hardwareModel':'ES240',
   'hardwareName':'Falcon',
   'deviceType': 'GATEWAYSTB',
   'gateway': true,
   'gatewayIPAddress': '192.168.1.20',
   'manufacturer':'Sky',
   'modelNumber':'Q112.000.21.00L (53wk8j8)',
   'serialNumber':'0627086857 2',
   'versionNumber':'32B12D',
   'bouquet': 4101,
   'subbouquet': 9,
   'wakeReason': 'ECO',
   'systemUptime': 28346,
   'hdrCapable': true,
   'uhdCapable': true,
   'presentLocalTimeOffset': 3600,
   'utc': 1660486276,
   'futureLocalTimeOffset': 0,
   'futureTransitionUtc': 1667091599
}
```

### Get device information (JSON)

```
device = self.client.get_device_information().as_json()
```

Will return a JSON structure such as below for device information:

```
{
   "__type__":"__device__",
   "attributes":{
      "ASVersion":"Q112.000.21.00-AS_asdev",
      "IPAddress":"192.168.x.xx",
      "countryCode":"GBR",
      "usedCountryCode":"GBR",
      "hardwareModel":"ES240",
      "hardwareName":"Falcon",
      "deviceType": "GATEWAYSTB",
      "gateway": true,
      "gatewayIPAddress": "192.168.1.20",
      "manufacturer":"Sky",
      "modelNumber":"Q112.000.21.00L (53wk8j8)",
      "serialNumber":"##########",
      "versionNumber":"32B12D",
      "bouquet": 4101,
      "subbouquet": 9,
      "wakeReason": "ECO",
      "systemUptime": 28346,
      "hdrCapable": true,
      "uhdCapable": true,
      "presentLocalTimeOffset": 3600,
      "utc": 1660486276,
      "futureLocalTimeOffset": 0,
      "futureTransitionUtc": 1667091599
   }
}

```

### Decode device information (JSON)

```
from pyskyqremote.media import device_decoder
device = device_decoder(device_json)
```

Will decode the JSON structure to a python object.


### Get power status

```
status = self.client.power_status()
```

Will return "POWERED OFF", "STANDBY or "ON"

### Get current state

```
status = self.client.get_current_state()
```

Will return an object such as below with current transport status information (`state` is the same as the previously returned single attribute):
```
{
    'CurrentTransportState':'PLAYING',
    'CurrentTransportStatus':'OK',
    'CurrentSpeed':'1',
    'state':'PLAYING'
}
```

### Get current state (JSON)

```
status = self.client.get_current_state().as_json()
```

Will return an object such as below with current transport status information (`state` is the same as the previously returned single attribute):
```
{
   "__type__":"__transportinfo__",
   "attributes":{
      "CurrentTransportState":"PLAYING",
      "CurrentTransportStatus":"OK",
      "CurrentSpeed":"1",
      "state":"PLAYING"
}
```

### Get the active application

```
app = self.client.get_active_application()
```

Will return an object such as below for the running application name:

```
{
   'appid':'com.bskyb.beehive',
   'title':'Beehive Bedlam'
}
```

### Get the active application (JSON)

```
app = self.client.get_active_application().as_json()
```

Will return a JSON structure such as below for the running application name:

```
{
   "__type__":"__app__",
   "attributes":{
      "appid":"com.bskyb.beehive",
      "title":"Beehive Bedlam"
   }
}
```

### Get current media

```
app = self.client.get_current_media()
```

Will return an object such as below for live programme:

```
{
   'channel':'Sky Comedy HD',
   'channelno':'153',
   'image_url':'https://d2n0069hmnqmmx.cloudfront.net/epgdata/1.0/newchanlogos/600/600/skychb1143.png',
   'sid':1143,
   'pvrid':None,
   'live':True
}
```
or for recording

```
{
   'channel':None,
   'channelno':'None',
   'image_url':None,
   'sid':None,
   'pvrid':'P12345ABC'
   'live':False
}
```
### Get current media (JSON)

```
media = self.client.get_current_media().as_json()
```

Will return a JSON structure such as below for live programme:

```
{
   "__type__":"__media__",
   "attributes":{
      "channel":"BBC One South",
      "channelno":"101",
      "image_url":"https://d2n0069hmnqmmx.cloudfront.net/epgdata/1.0/newchanlogos/600/600/skychb2153.png",
      "sid":2153,
      "pvrid":null,
      "live":true
   }
}

```
or for recording

```
{
   "__type__":"__media__",
   "attributes":{
      "channel":null,
      "channelno":null,
      "image_url":null,
      "sid":null,
      "pvrid":"P12345ABC",
      "live":false
   }
}
```

### Decode current media (JSON)

```
from pyskyqremote.media import media_decoder
media = media_decoder(media_json)
```

Will decode the JSON structure to a python object.


### Get EPG information

```
epg = self.client.get_epg_data(sid, epgDate, days)
```

Will return an object with an array of events including the specified number of days. Defaults to 2 days.

```
{
   'sid':2153,
   'channelno':'101',
   'channelname':'BBC One South',
   'channelimage_url':'https://d2n0069hmnqmmx.cloudfront.net/epgdata/1.0/newchanlogos/600/600/skychb2153.png',
   'programmes':[
      {
         'progammeuuid':'57a11caf-1ebd-4c01-a40b-7fdfe5c5fad0',
         'starttime':datetime.datetime(2020,4,16,21,50,tzinfo=datetime.timezone.utc),
         'endtime':datetime.datetime(2020,4,16,22,50,tzinfo=datetime.timezone.utc),
         'title':'New: Tonight Show Starring. Jimmy Fallon',
         'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
         'season':7,
         'episode':119,
         'image_url':'https://images.metadata.sky.com/pd-image/57a11caf-1ebd-4c01-a40b-7fdfe5c5fad0/16-9',
         'channelname':'BBC One South',
         'status':'LIVE',
         'eid':'E4b8-19b'
      },
      {
         'progammeuuid':'d2d67048-673a-4ea8-8a32-3ad386e306d2',
         'starttime':datetime.datetime(2020,4,16,22,50,tzinfo=datetime.timezone.utc),
         'endtime':datetime.datetime(2020,4,16,23,50,tzinfo=datetime.timezone.utc),
         'title':'New: Late Late Show With...',
         'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
         'season':2020,
         'episode':89,
         'image_url':'https://images.metadata.sky.com/pd-image/d2d67048-673a-4ea8-8a32-3ad386e306d2/16-9',
         'channelname':'BBC One South',
         'status':'LIVE',
         'eid':'E4b8-19b'
      },
      {...}
   ]
}
```
### Get EPG information (JSON)

```
epg = self.client.get_epg_data(sid, epgDate, days).as_json()
```

Will return a JSON structure with an array of events including the specified number of days. Defaults to 2 days.

```
{
   "__type__":"__channel__",
   "attributes":{
      "sid":2153,
      "channelno":"101",
      "channelname":"BBC One South",
      "channelimage_url":"https://d2n0069hmnqmmx.cloudfront.net/epgdata/1.0/newchanlogos/600/600/skychb2153.png"
   },
   "programmes":[
      {
         "__type__":"__programme__",
         "attributes":{
            "programmeuuid":"62ad0457-1a6a-4b45-9ef7-6e144639d734",
            "starttime":"2020-04-27T21:45:00+0000",
            "endtime":"2020-04-27T22:10:00+0000",
            "title":"Man Like Mobeen",
            "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
            "season":3,
            "episode":5,
            "image_url":"https://images.metadata.sky.com/pd-image/62ad0457-1a6a-4b45-9ef7-6e144639d734/16-9",
            "channelname":"BBC One South",
            "status":"LIVE",
            "eid":"E4b8-19b"
        }
      },
      {
         "__type__":"__programme__",
         "attributes":{
            "programmeuuid":"a975bdeb-c19b-4de2-9557-c6d2757bdae7",
            "starttime":"2020-04-27T22:10:00+0000",
            "endtime":"2020-04-27T22:50:00+0000",
            "title":"Have I Got A Bit More News For You",
            "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
            "season":59,
            "episode":4,
            "image_url":"https://images.metadata.sky.com/pd-image/a975bdeb-c19b-4de2-9557-c6d2757bdae7/16-9",
            "channelname":"BBC One South"
            "status":"LIVE",
            "eid":"E4b8-19b"
         }
      },
      {...},
   ]
}
```

### Decode EPG information (JSON)

```
from pyskyqremote.channel import channel_decoder
channel = channel_decoder(channel_json)
```

Will decode the JSON structure to a python object.

### Get programme at a point in time on a day

Note that at the end of a day, the programme may appear on the next day's schedule.

```
programme = self.client.get_programme_from_epg(sid, epgDate, queryDate)
```

Will return an object such as below:

```
{
   'progammeuuid':'9fbdcefe-312c-4681-b996-00637e85313a',
   'starttime':datetime.datetime(2020,4,17,8,30,tzinfo=datetime.timezone.utc),
   'endtime':datetime.datetime(2020,4,17,9,0,tzinfo=datetime.timezone.utc),
   'title':'Parks And Recreation',
   'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
   'season':4,
   'episode':5,
   'image_url':'https://images.metadata.sky.com/pd-image/9fbdcefe-312c-4681-b996-00637e85313a/16-9',
   'channelname':'Channel 5 HD',
   'status':'LIVE',
   'eid':'E4b8-19b'
}
```
### Get programme at a point in time on a day (JSON)

Note that at the end of a day, the programme may appear on the next day's schedule.

```
programme = self.client.get_programme_from_epg(sid, epgDate, queryDate).as_json()
```

Will return a JSON structure such as below:

```
{
   "__type__":"__programme__",
   "attributes":{
      "programmeuuid":"e11d9e93-0eec-4855-88f5-6ade9946d5dd",
      "starttime":"2020-04-28T21:00:00+0000",
      "endtime":"2020-04-28T21:30:00+0000",
      "title":"BBC News at Ten",
      "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
      "season":null,
      "episode":null,
      "image_url":"https://images.metadata.sky.com/pd-image/e11d9e93-0eec-4855-88f5-6ade9946d5dd/16-9",
      "channelname":"BBC ONE HD",
      "status":"LIVE",
      "eid':"E4b8-19b"
   }
}
```
### Decode programme information (JSON)

```
from pyskyqremote.programme import programme_decoder
programme = programme_decoder(programmeJSON)
```

### Get current live TV programme on a channel

```
currentTV = self.client.get_current_live_tv_programme(sid)
```

Will return an object such as below:

```
{
   'progammeuuid':'9fbdcefe-312c-4681-b996-00637e85313a',
   'starttime':datetime.datetime(2020,4,17,8,30,tzinfo=datetime.timezone.utc),
   'endtime':datetime.datetime(2020,4,17,9,0,tzinfo=datetime.timezone.utc),
   'title':'Parks And Recreation',
   'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
   'season':4,
   'episode':5,
   'image_url':'https://images.metadata.sky.com/pd-image/9fbdcefe-312c-4681-b996-00637e85313a/16-9',
   'channelname':'Channel 5 HD',
   'status':'LIVE',
   'eid':'E4b8-19b'
}
```
### Get current live TV programme on a channel (JSON)

```
currentTV = self.client.get_current_live_tv_programme(sid).as_json()
```

Will return a JSON structure such as below:

```
{
   "__type__":"__programme__",
   "attributes":{
      "programmeuuid":"e11d9e93-0eec-4855-88f5-6ade9946d5dd",
      "starttime":"2020-04-28T21:00:00+0000",
      "endtime":"2020-04-28T21:30:00+0000",
      "title":"BBC News at Ten",
      "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
      "season":null,
      "episode":null,
      "image_url":"https://images.metadata.sky.com/pd-image/e11d9e93-0eec-4855-88f5-6ade9946d5dd/16-9",
      "channelname":"BBC ONE HD",
      "status":"LIVE",
      "eid":"E4b8-19b"
  }
}
```

### Get recordings

```
recordings = self.client.get_recordings(<status="all">, <limit=1000>, <offset=0>)
```

Will return an object such as below for the number of recordings specified by limit with the specified offset and with the provided status:

```
{
   'programmes':[
      {
        'programmeuuid':'54bfc205-c56e-4583-b03f-59c31f97f8c7',
        'starttime':datetime.datetime(2020,4,17,8,30,tzinfo=datetime.timezone.utc),
        'endtime':datetime.datetime(2020,4,17,9,30,tzinfo=datetime.timezone.utc),
        'title':'New: Batwoman',
        'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
        'summary': 'Lorum ipsum...',
        'season':1,
        'episode':19,
        'image_url':'https://images.metadata.sky.com/pd-image/54bfc205-c56e-4583-b03f-59c31f97f8c7/16-9',
        'channelname':'E4 HD',
        'status':'RECORDED',
        'deletetime': '2020-09-02T20:00:59Z',
        'failurereason': None,
        'source': 'LIVE',
        'pvrid':'P29014192',
        'eid':'E869-67b1',
      },
      {
        'programmeuuid':'af9ecd2c-5026-4050-9c15-37598fe26713',
        'starttime':null,
        'endtime':'null,
        'title':'Home and Away',
        'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
        'summary': 'Lorum ipsum...',
        'season':35,
        'episode':4,
        'image_url':'https://images.metadata.sky.com/pd-image/af9ecd2c-5026-4050-9c15-37598fe26713/16-9',
        'channelname':'Channel 5 HD',
        'status':'SCHEDULED',
        'deletetime': None,
        'failurereason': None,
        'source': 'LIVE',
        'pvrid':'P29014192',
        'eid':'E869-67b1'
      },
      {
        'programmeuuid':'575736fd-0719-4249-88cc-babd6e232bfa',
        'starttime':datetime.datetime(2020,4,17,8,30,tzinfo=datetime.timezone.utc),
        'endtime':datetime.datetime(2020,4,17,9,30,tzinfo=datetime.timezone.utc),
        'title':'Lorraine',
        'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
        'summary': 'Lorum ipsum...',
        'season':35,
        'episode':4,
        'image_url':'https://images.metadata.sky.com/pd-image/575736fd-0719-4249-88cc-babd6e232bfa/16-9',
        'channelname':'ITV HD',
        'status':'PART REC',
        'deletetime': None,
        'failurereason': 'Start Missed',
        'source': 'VOD',
        'pvrid':'P29014192',
        'eid':'E869-67b1'
      },
      {…}
      }
   ]
}
```

### Get recordings (JSON)

```
recordings = self.client.get_recordings(<status="all">, <limit=1000>, <offset=0>).as_json()
```

Will return an object such as below for the number of recordings specified by limit with the specified offset and with the provided status:

```
{
   "__type__":"__recordings__",
   "attributes":{

   },
   "recordings":[
      {
         "__type__":"__recording__",
         "attributes":{
            "programmeuuid":"54bfc205-c56e-4583-b03f-59c31f97f8c7",
            "starttime":"2020-08-02T19:58:00+0000",
            "endtime":"2020-08-02T21:01:59+0000",
            "title":"New: Batwoman",
            "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
            "summary":"Lorum ipsum...",
            "season":1,
            "episode":19,
            "image_url":"https://images.metadata.sky.com/pd-image/54bfc205-c56e-4583-b03f-59c31f97f8c7/16-9",
            "channelname":"E4 HD",
            "status":"RECORDED",
            "deletetime":"2020-09-02T20:00:59+0000",
            "failurereason":"null",
            "source":"LIVE",
            "pvrid":"P29014192",
            "eid":"E869-67b1"
        }
      },
      {
         "__type__":"__recording__",
         "attributes":{
            "programmeuuid":"af9ecd2c-5026-4050-9c15-37598fe26713",
            "starttime":"null",
            "endtime":"null",
            "title":"Home and Away",
            "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
            "summary":"Lorum ipsum...",
            "season":35,
            "episode":4,
            "image_url":"https://images.metadata.sky.com/pd-image/af9ecd2c-5026-4050-9c15-37598fe26713/16-9",
            "channelname":"Channel 5 HD",
            "status":"SCHEDULED",
            "deletetime":"null",
            "failurereason":"null",
            "source":"LIVE",
            "pvrid":"P29014192",
            "eid":"E869-67b1"
     },
     {
        "__type__":"__recording__",
        "attributes":{
           "programmeuuid":"af9ecd2c-5026-4050-9c15-37598fe26713",
           "starttime":"2020-08-02T19:58:00+0000",
           "endtime":"2020-08-02T21:01:59+0000",
           "title":"Home and Away",
           "synopsis":"1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.",
           "summary":"Lorum ipsum...",
           "season":35,
           "episode":4,
           "image_url":"https://images.metadata.sky.com/pd-image/af9ecd2c-5026-4050-9c15-37598fe26713/16-9",
           "channelname":"Channel 5 HD",
           "status":"PART REC",
           "deletetime":"null",
           "failurereason":"Start Missed",
           "source":"VOD",
           "pvrid":"P29014192",
           "eid":"E869-67b1"
    },
      {...}
      }
   ]
}
```

### Get recording

```
recording = self.client.getRecording(pvrid)
```

Will return an object such as below:

```
{
   'progammeuuid':'9fbdcefe-312c-4681-b996-00637e85313a',
   'starttime':datetime.datetime(2020,4,17,8,30,tzinfo=datetime.timezone.utc),
   'endtime':datetime.datetime(2020,4,17,9,0,tzinfo=datetime.timezone.utc),
   'channelname':'ITV HD',
   'title':'Van Der Valk',
   'synopsis':'1/6. Following ambulance staff during a historical period of uncertainty for the NHS. Contains some strong language and upsetting scenes.',
   'season':4,
   'episode':5,
   'image_url':'https://images.metadata.sky.com/pd-image/ddcd727f-487f-4558-8365-7bed4fe41c87/16-9',
   'status':'RECORDED',
   'deletetime': None,
   'failurereason': None,
   'pvrid':'P29014192',
   'eid':'E869-67b1'
}
```
### Get recording (JSON)

```
recording = self.client.getRecording(pvrid).as_json()
```

Will return an object such as below:

```
{
   "__type__":"__recording__",
   "attributes":{
      "programmeuuid":"e11d9e93-0eec-4855-88f5-6ade9946d5dd",
      "starttime":"2020-04-28T21:00:00+0000",

CHANGELOG.md
# Changelog

## v0.3.26 (2023/08/29)
### Enhancements
- [Add synopsis to programme/recording information](https://github.com/RogerSelwyn/skyq_remote/commit/a42e64c196d7b22d0a9168f05c9692562327e60b) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.26](https://github.com/RogerSelwyn/skyq_remote/commit/1e219dfaebb7a930861ee79b8a49745659f7817a) - @RogerSelwyn

### Documentation
- [Update README.md](https://github.com/RogerSelwyn/skyq_remote/commit/c0dd51a4528f4175fbfe079b33d790620398c6fc) - @RogerSelwyn

## v0.3.25 (2023/04/01)
### Enhancements
- [Make datetimes timezone aware (UTC)](https://github.com/RogerSelwyn/skyq_remote/commit/e48fe78d88eab857cedd8ddc83edaa52010e1879) - @RogerSelwyn
### Maintenance
- [Bump to 0.3.25](https://github.com/RogerSelwyn/skyq_remote/commit/e8ab8934cc6640df9926ecfde59af0b3ebc92526) - @RogerSelwyn

## v0.3.24 (2023/01/18)
### Fixes
- [Fix start/end time being returned as 0 when not present](https://github.com/RogerSelwyn/skyq_remote/commit/9cfe70c7afb402682098a0a83cf815ac14924422) - @RogerSelwyn

### Maintenance
- [Update README.md](https://github.com/RogerSelwyn/skyq_remote/commit/6212deb092a3d836287d2c2a515df2791a15183f) - @RogerSelwyn
- [Bump to 0.3.24](https://github.com/RogerSelwyn/skyq_remote/commit/993f23a5b63d4d7311d4c786ae5106549e726ea5) - @RogerSelwyn

## v0.3.23 (2022/12/29)
### Maintenance
- [Remove gatewayWakeReason](https://github.com/RogerSelwyn/skyq_remote/commit/331b5d88fff97eee3e1c9277d3ba27a1a8416f9f) - @RogerSelwyn
- [Bump 0.3.23](https://github.com/RogerSelwyn/skyq_remote/commit/3479b541d27b4639b536ba84f7dee918fb21057c) - @RogerSelwyn

## v0.3.22 (2022/12/26)
### Enhancements
- [Add gatewayIP and gatewayWakeReason](https://github.com/RogerSelwyn/skyq_remote/commit/d9981a865036576c57387886923e686968f4c42c) - @RogerSelwyn
### Maintenance
- [Update README.md](https://github.com/RogerSelwyn/skyq_remote/commit/862ec801063a1708a39e438514fa35ff42e67880) - @RogerSelwyn
- [Bump to 0.3.22](https://github.com/RogerSelwyn/skyq_remote/commit/fab7f4d3bf3208034233a37db669a93bad15bed3) - @RogerSelwyn

## v0.3.21 (2022/11/25)
### Maintenance
- [Sourcery recommended code improvements](https://github.com/RogerSelwyn/skyq_remote/commit/a4bd5600a29a443bed27b1e9aaa1587be939dff9) - @RogerSelwyn
- [Bump to 0.3.21](https://github.com/RogerSelwyn/skyq_remote/commit/4e8dadf361c698e3ca3005d223c723939d47bb31) - @RogerSelwyn

## v0.3.20 (2022/11/10)
### Enhancements
- [Add summary and source to recording](https://github.com/RogerSelwyn/skyq_remote/commit/f528bb65e743909562ed51c1d73d1dd1ef108a26) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.20](https://github.com/RogerSelwyn/skyq_remote/commit/cc88abae21fa337e09cefacb9a58e4fc17272f5d) - @RogerSelwyn

## v0.3.19 (2022/11/09)
### Fixes
- [Fix issue with live recording](https://github.com/RogerSelwyn/skyq_remote/commit/54b605b3713c0b278deae80987d22a7095aa521e) - @RogerSelwyn

### Maintenance
- [Simplify methods](https://github.com/RogerSelwyn/skyq_remote/commit/7cd383b606d2cf3300f998f8e3b6e75690ba7d30) - @RogerSelwyn
- [Bump to 0.3.19](https://github.com/RogerSelwyn/skyq_remote/commit/780e50c87b4464d1b6dd5d3d752c2befc57fe2b9) - @RogerSelwyn

## v0.3.18 (2022/11/09)
### Enhancements
- [Make start/end time more accurate and break recordings class our from programmes. Also add delete time and failure reason](https://github.com/RogerSelwyn/skyq_remote/commit/3994991e01da8e8b4ed169438f5f5b8c8dbe8480) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.18](https://github.com/RogerSelwyn/skyq_remote/commit/c373c3825cdced62e84ff3ea6121396a853fa1ed) - @RogerSelwyn

## v0.3.17 (2022/10/25)
### Maintenance
- [Add logging](https://github.com/RogerSelwyn/skyq_remote/commit/a19f2c9453949e61bdfc5a1fd26a8c12b7bd246e) - @RogerSelwyn
- [Reduce calls](https://github.com/RogerSelwyn/skyq_remote/commit/42bb7620c07527266ae3d70964c99f0edcfa1150) - @RogerSelwyn
- [Tweaks to support recording info in HA](https://github.com/RogerSelwyn/skyq_remote/commit/9fc7742bdc68af4f5ebe814233d1c37a1c85215c) - @RogerSelwyn
- [Bump to 0.3.17](https://github.com/RogerSelwyn/skyq_remote/commit/ba081ce476c354955301db8865faab6a5114a256) - @RogerSelwyn

## v0.3.16 (2022/08/14)
### Enhancements
- [Add Time info to device information](https://github.com/RogerSelwyn/skyq_remote/commit/9d729bf63e7bcb332c30877eea7b7917b3507eb4) - @RogerSelwyn
### Maintenance
- [Bump to 0.3.16](https://github.com/RogerSelwyn/skyq_remote/commit/18de4a34f2c7d4aab2e9f68e44bbfdeeff23b77c) - @RogerSelwyn

## v0.3.15 (2022/08/08)
### Enhancements
- [Increase size of image](https://github.com/RogerSelwyn/skyq_remote/commit/4799df3245a653306fdd8a1945ca4070e63d95a7) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.15](https://github.com/RogerSelwyn/skyq_remote/commit/71245b94f93e937e848dae8d68a5bdba25d989f2) - @RogerSelwyn

## v0.3.14 (2022/07/30)
### Fixes
- [Fix issue with hostname in SOAP calls](https://github.com/RogerSelwyn/skyq_remote/commit/30a9e2eff3a715c16d51c11717ebde11544777b9) - @RogerSelwyn
- [Fix transport info](https://github.com/RogerSelwyn/skyq_remote/commit/23d14d5485020395e1f194d5c1ab74fdc583a444) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.13](https://github.com/RogerSelwyn/skyq_remote/commit/79a3ab2d185d8650000b814dc6d0ae1b79f57daf) - @RogerSelwyn
- [Bump to 0.3.14](https://github.com/RogerSelwyn/skyq_remote/commit/47d662f9939ffb2d35905c1e2fa4d34b97042cd4) - @RogerSelwyn

## v0.3.12 (2022/05/21)
### Fixes
- [Fix issue with restart during eco mode](https://github.com/RogerSelwyn/skyq_remote/commit/367c25791d2026267ffe100c789a9f15779673fa) - @RogerSelwyn

### Maintenance
- [Bump to 0.3.12](https://github.com/RogerSelwyn/skyq_remote/commit/a571c5eb3c5e86c50159282532a9731e3276eea8) - @RogerSelwyn

## v0.3.11 (2022/05/20)
### Fixes
- [Fix errors displayed during ECO mode](https://github.com/RogerSelwyn/skyq_remote/commit/4926f431c553ff979d6f296e9c043d2ca2408c26) - @RogerSelwyn

### Maintenance

pyskyqremote/const.py
"""Constants for pyskyqremote."""
CONST_DATE_FORMAT = "%Y-%m-%dT%H:%M:%S%z"
# SOAP/UPnP Constants
SKY_PLAY_URN = "urn:nds-com:serviceId:SkyPlay"
SKYCONTROL = "SkyControl"
SOAP_ACTION = '"urn:schemas-nds-com:service:SkyPlay:2#{0}"'
SOAP_CONTROL_BASE_URL = "http://{0}:49153{1}"
SOAP_DESCRIPTION_BASE_URL = "http://{0}:49153/description{1}.xml"
SOAP_PAYLOAD = """<s:Envelope xmlns:s='http://schemas.xmlsoap.org/soap/envelope/' s:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/'>  # pylint: disable=line-too-long
    <s:Body>
        <u:{0} xmlns:u="urn:schemas-nds-com:service:SkyPlay:2">
            <InstanceID>0</InstanceID>
        </u:{0}>
    </s:Body>
</s:Envelope>"""
SOAP_RESPONSE = "u:{0}Response"
SOAP_USER_AGENT = "SKYPLUS_skyplus"
UPNP_GET_MEDIA_INFO = "GetMediaInfo"
UPNP_GET_TRANSPORT_INFO = "GetTransportInfo"

# WebSocket Constants
WS_BASE_URL = "ws://{0}:9006/as/{1}"
WS_CURRENT_APPS = "apps/status"

# REST Constants
REST_BASE_URL = "http://{0}:{1}/as/{2}"
REST_CHANNEL_LIST = "services/{0}/{1}"
REST_FAVOURITES = "services/favourites"
REST_QUOTA_DETAILS = "pvr/storage"
REST_BOOK_RECORDING = "pvr/action/bookrecording?eid={0}"
REST_BOOK_PPVRECORDING = "pvr/action/bookppvrecording?eid={0}&offerref={1}"
REST_BOOK_SERIES_RECORDING = "pvr/action/bookseriesrecording?eid={0}"
REST_RECORDING_DETAILS = "pvr/details/{0}"
REST_RECORDINGS_LIST = "pvr/?limit={0}&offset={1}"
REST_RECORDING_KEEP = "pvr/action/keep?pvrid={0}"
REST_RECORDING_UNKEEP = "pvr/action/unkeep?pvrid={0}"
REST_RECORDING_LOCK = "pvr/action/lock?pvrid={0}"
REST_RECORDING_UNLOCK = "pvr/action/unlock?pvrid={0}"
REST_RECORDING_DELETE = "pvr/action/delete?pvrid={0}"
REST_RECORDING_UNDELETE = "pvr/action/undelete?pvrid={0}"
REST_RECORDING_ERASE = "pvr/action/erase?pvrid={0}"
REST_RECORDING_ERASE_ALL = "pvr"
REST_RECORDING_SET_LAST_PLAYED_POSITION = (
    "pvr/action/setlastplayedposition?pos={0}&pvrid={1}"
)
REST_SERIES_LINK = "pvr/action/serieslink?pvrid={0}"
REST_SERIES_UNLINK = "pvr/action/seriesunlink?pvrid={0}"
REST_SYSTEM_TIME = "system/time"
REST_PATH_SYSTEMINFO = "system/information"
REST_PATH_DEVICEINFO = "system/deviceinformation"
REST_PATH_APPS = "apps"
REST_GET = "get"
REST_POST = "post"
REST_DELETE = "delete"

# Sky specific constants
CURRENT_URI = "CurrentURI"
CURRENT_TRANSPORT_STATE = "CurrentTransportState"
CURRENT_TRANSPORT_STATUS = "CurrentTransportStatus"
CURRENT_SPEED = "CurrentSpeed"
DEFAULT_TRANSPORT_STATE = "OK"
DEFAULT_TRANSPORT_SPEED = 1
APP_STATUS_VISIBLE = "VISIBLE"

PVR = "pvr"
XSI = "xsi"

SKY_STATE_NOMEDIA = "NO_MEDIA_PRESENT"
SKY_STATE_OFF = "POWERED OFF"
SKY_STATE_ON = "ON"
SKY_STATE_PLAYING = "PLAYING"
SKY_STATE_PAUSED = "PAUSED_PLAYBACK"
SKY_STATE_STANDBY = "STANDBY"
SKY_STATE_STOPPED = "STOPPED"
SKY_STATE_TRANSITIONING = "TRANSITIONING"
SKY_STATE_UNSUPPORTED = "UNSUPPORTED"
SKY_STATUS_LIVE = "LIVE"

APP_EPG = "com.bskyb.epgui"

COMMANDS = {
    "power": 0,
    "select": 1,
    "backup": 2,
    "dismiss": 2,
    "channelup": 6,
    "channeldown": 7,
    "interactive": 8,
    "sidebar": 8,
    "help": 9,
    "services": 10,
    "search": 10,
    "tvguide": 11,
    "home": 11,
    "i": 14,
    "text": 15,
    "up": 16,
    "down": 17,
    "left": 18,
    "right": 19,
    "red": 32,
    "green": 33,
    "yellow": 34,
    "blue": 35,
    "0": 48,
    "1": 49,
    "2": 50,
    "3": 51,
    "4": 52,
    "5": 53,
    "6": 54,
    "7": 55,
    "8": 56,
    "9": 57,
    "play": 64,
    "pause": 65,
    "stop": 66,
    "record": 67,
    "fastforward": 69,
    "rewind": 71,
    "boxoffice": 240,
    "sky": 241,
}

KNOWN_COUNTRIES = {
    "DEU": "DEU",
    "GBR": "GBR",
    "IRL": "GBR",
    "ITA": "ITA",
}
TERRITORIES = {"DEU": "DE", "GBR": "GB", "ITA": "IT"}
URL_PREFIX = {"DEU": "de.", "GBR": "", "ITA": "it."}

# Random set of other constants
EPG_ERROR_PAST_END = "past end of epg"
EPG_ERROR_NO_DATA = "no epg data found"

RESPONSE_OK = 200
CONNECT_TIMEOUT = 1000
EPG_TIMEOUT = 60
HTTP_TIMEOUT = 6
SOAP_TIMEOUT = 2

AUDIO = "audio"
VIDEO = "video"

ALLRECORDINGS = "all"

REST_GET = "get"
REST_POST = "post"
REST_DELETE = "DELETE"

DEVICE_GATEWAYSTB = "GATEWAYSTB"
DEVICE_IPSETTOPBOX = "IPSETTOPBOX"
DEVICE_MULTIROOMSTB = "MULTIROOMSTB"
DEVICE_TV = "TV"
UNSUPPORTED_DEVICES = [DEVICE_IPSETTOPBOX, DEVICE_TV]

SCHEDULE_URL = "http://atlantis.epgsky.com/as/schedule/{1}/{0}"
LIVE_IMAGE_URL = (
    "https://{1}imageservice.sky.com/pd-image/{0}/16-9/1024?territory={2}"
    + "&provider=SKY&proposition=SKYQ"
)
PVR_IMAGE_URL = LIVE_IMAGE_URL
CHANNEL_IMAGE_URL = (
    "https://{2}imageservice.sky.com/logo/skychb_{0}{1}/600/600?territory={3}"
    + "&provider=SKY&proposition=SKYQ"
)
